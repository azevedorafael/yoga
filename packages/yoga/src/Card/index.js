import Card from './web/Card';
import CardFooter from './web/CardFooter';
import CardHeader from './web/CardHeader';
import Plan from './web/Plan';

Card.Footer = CardFooter;
Card.Header = CardHeader;
Card.Plan = Plan;

export default Card;
