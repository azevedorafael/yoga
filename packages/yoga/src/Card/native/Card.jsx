import React from 'react';
import styled from 'styled-components';

const CardStyled = styled.View(
  ({
    theme: {
      yoga: {
        components: { card },
      },
    },
  }) => `
    padding:
      ${card.padding.top}px 
      ${card.padding.right}px 
      ${card.padding.bottom}px 
      ${card.padding.left}px;

    border-radius: ${card.radii}px;
    background-color: ${card.backgroundColor};
    box-shadow: ${card.elevation};
    elevation: 4;
`,
);

const Card = ({ ...rest }) => <CardStyled {...rest} />;

Card.displayName = 'Card';

export default Card;
