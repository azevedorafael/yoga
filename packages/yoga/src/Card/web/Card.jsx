import React from 'react';
import styled from 'styled-components';

const CardStyled = styled.div(
  ({
    theme: {
      components: { card },
    },
  }) => `
    display: flex;
    flex-direction: column;
    justify-content: flex-start;

    width: 100%;
    padding: 
      ${card.padding.top}px 
      ${card.padding.right}px 
      ${card.padding.bottom}px 
      ${card.padding.left}px;

    border-radius: ${card.radii}px;
    background-color: ${card.background};
    box-shadow: ${card.elevation};
`,
);

const Card = ({ ...rest }) => <CardStyled {...rest} />;

Card.displayName = 'Card';

export default Card;
