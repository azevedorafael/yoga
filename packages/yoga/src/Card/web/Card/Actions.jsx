import styled from 'styled-components';
import { arrayOf, oneOfType } from 'prop-types';

import Button from '../../../Button';
import { typeOf } from '../../../shared/propTypes';

const Actions = styled.footer``;

Actions.propTypes = {
  children: oneOfType([
    oneOfType([
      typeOf(Button),
      typeOf(Button.Text),
      typeOf(Button.Outline),
      typeOf(Button.Link),
    ]),
    arrayOf(
      oneOfType([
        typeOf(Button),
        typeOf(Button.Text),
        typeOf(Button.Outline),
        typeOf(Button.Link),
      ]),
    ),
  ]).isRequired,
};

Actions.displayName = 'Card.Actions';

export default Actions;
