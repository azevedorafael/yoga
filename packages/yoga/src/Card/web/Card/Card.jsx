import React from 'react';
import styled from 'styled-components';
import { string, shape, oneOf, node } from 'prop-types';

import Text from '../../../Text';

const CardStyled = styled.section`
  ${({
    variant,
    variantIntensity,
    theme: {
      yoga: {
        colors: { [variant]: color = {} },
        components: { card },
      },
    },
  }) => `
    padding: 
      ${card.padding.top}px 
      ${card.padding.right}px 
      ${card.padding.bottom}px 
      ${card.padding.left}px;

    border-radius: ${card.radii}px;
    background-color: ${
      variant ? color[variantIntensity] : card.backgroundColor
    };
    box-shadow: ${card.elevation};
  `}
`;

const Ribbon = styled(Text.Tiny)`
  ${({
    variant,
    variantIntensity,
    theme: {
      yoga: {
        components: { card },
        spacing,
        colors: { gray, [variant]: color = {}, white },
      },
    },
  }) => `
    display: inline-block;
    top: ${card.padding.top}px;
    left: 0;

    margin-left: -${card.padding.left}px;
    margin-bottom: ${card.padding.bottom}px;

    padding: 
      ${spacing.xxsmall}px
      ${spacing.small}px
      ${spacing.xxsmall}px
      ${spacing.medium}px;
    
    border-top-right-radius: ${card.ribbon.radius}px;
    border-bottom-right-radius: ${card.ribbon.radius}px;

    background-color: ${
      variant
        ? color[typeof variantIntensity === 'number' ? variantIntensity : 2]
        : gray.i10
    };

    color: ${variant ? white : gray.i80};
  `}
`;

const Card = ({ ribbon, children, ...rest }) => (
  <CardStyled {...rest}>
    {Object.keys(ribbon).length > 0 && (
      <Ribbon
        variant={ribbon.variant}
        variantIntensity={ribbon.variantIntensity}
        as="span"
      >
        {ribbon.text}
      </Ribbon>
    )}
    {children}
  </CardStyled>
);

Card.propTypes = {
  /** text: the content inside the Card Ribbon
   * variant: style the ribbon following the theme (primary, secondary, tertiary)
   * variantIntensity: ribbon variant color intensity (0, 1, 2) */
  ribbon: shape({
    text: string,
    variant: oneOf(['', 'primary', 'secondary']),
    variantIntensity: oneOf([0, 1, 2]),
  }),
  children: node,
  /** style the card following the theme (primary, secondary, tertiary) */
  variant: oneOf(['', 'primary', 'secondary']),
  /** intensity of variant color (0, 1, 2) */
  variantIntensity: oneOf([0, 1, 2]),
};

Card.defaultProps = {
  ribbon: {},
  children: null,
  variant: '',
  variantIntensity: 2,
};

Card.displayName = 'Card';

export default Card;
