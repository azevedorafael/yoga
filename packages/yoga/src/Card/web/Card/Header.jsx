import styled from 'styled-components';

const Header = styled.header``;

Header.displayName = 'Card.Header';

export default Header;
