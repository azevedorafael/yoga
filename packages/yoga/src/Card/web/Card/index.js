import Card from './Card';
import Header from './Header';
import Content from './Content';
import Actions from './Actions';

Card.Header = Header;
Card.Content = Content;
Card.Actions = Actions;

export default Card;
