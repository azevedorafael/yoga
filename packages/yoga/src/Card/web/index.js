import Card from './Card';
import Footer from './Footer';
import Header from './Header';
import Plan from './Plan';

export { Card, Footer, Header, Plan };
