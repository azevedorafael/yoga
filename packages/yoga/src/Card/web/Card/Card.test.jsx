import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider, Button } from '../../..';
import Card from '.';

describe('<Card />', () => {
  describe('Snapshots', () => {
    it('should match snapshot with default Card', () => {
      const { container } = render(
        <ThemeProvider>
          <Card>
            <Card.Header>Hello World</Card.Header>
            <Card.Content>It&apos;s fine here</Card.Content>
            <Card.Actions>
              <Button>Action&apos;s children must be Button</Button>
            </Card.Actions>
          </Card>
        </ThemeProvider>,
      );
      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with variant Card', () => {
      const { container } = render(
        <ThemeProvider>
          <Card variant="primary">
            <Card.Header>Hello World</Card.Header>
            <Card.Content>It&apos;s fine here</Card.Content>
            <Card.Actions>
              <Button>Action&apos;s children must be Button</Button>
            </Card.Actions>
          </Card>
          <Card variant="secondary">
            <Card.Header>Hello World</Card.Header>
            <Card.Content>It&apos;s fine here</Card.Content>
            <Card.Actions>
              <Button>Action&apos;s children must be Button</Button>
            </Card.Actions>
          </Card>
          <Card variant="tertiary">
            <Card.Header>Hello World</Card.Header>
            <Card.Content>It&apos;s fine here</Card.Content>
            <Card.Actions>
              <Button>Action&apos;s children must be Button</Button>
            </Card.Actions>
          </Card>
        </ThemeProvider>,
      );
      expect(container).toMatchSnapshot();
    });
  });
});
