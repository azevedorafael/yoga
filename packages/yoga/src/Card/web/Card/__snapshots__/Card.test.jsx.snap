// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> Snapshots should match snapshot with default Card 1`] = `
.c1 {
  box-sizing: border-box;
  outline: none;
  -webkit-transition: all 0.2s;
  transition: all 0.2s;
  cursor: pointer;
  width: auto;
  height: 48px;
  padding-left: 24px;
  padding-right: 24px;
  background-color: #F46152;
  border: 2px solid #F46152;
  border-radius: 9999px;
  color: #FFFFFF;
  font-size: 14px;
  font-weight: 700;
  font-family: Open Sans;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
  line-height: 1;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c1:not([disabled]):hover,
.c1:not([disabled]):focus {
  box-shadow: 0 4px 8px rgba(244,97,82,0.45);
}

.c1:active {
  background-color: #F5A391;
  border-color: #F5A391;
  color: #FFFFFF;
}

.c1:disabled {
  background-color: #D3D3E2;
  border-color: #D3D3E2;
  color: #9898A7;
  cursor: not-allowed;
}

.c0 {
  padding: 16px  16px  16px  16px;
  border-radius: 8px;
  background-color: #FFFFFF;
  box-shadow: 0 2px 6px rgba(152,152,167,0.25);
}

<div>
  <section
    class="c0"
  >
    <header
      class=""
    >
      Hello World
    </header>
    <div
      class=""
    >
      It's fine here
    </div>
    <footer
      class=""
    >
      <button
        class="c1"
      >
        Action's children must be Button
      </button>
    </footer>
  </section>
</div>
`;
